File: src/foreign_if/spark/examples/scala/ALSDemo.scala
package test.scala;

import com.nec.frovedis.Jexrpc.FrovedisServer
import com.nec.frovedis.mllib.recommendation.ALS
import com.nec.frovedis.mllib.recommendation.MatrixFactorizationModel
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.mllib.recommendation.Rating

object ALSDemo {
  def main(args: Array[String]): Unit = {

    // -------- configurations --------
    val conf = new SparkConf().setAppName("ALSExample").setMaster("local[2]")
    val sc = new SparkContext(conf)

    // initializing Frovedis server with "personalized command", if provided in command line
    if(args.length != 0) FrovedisServer.initialize(args(0))

    // -------- data loading from sample libSVM file at Spark side--------
    var rating = sc.textFile("./input/rating.txt")
    val s_data = rating.map(_.split(',') match { case Array(user, item, rate) =>
                             Rating(user.toInt, item.toInt, rate.toDouble)
                  })

    // -------- train with default training params --------
    val rank = 4
    val m1 = ALS.trainImplicit(s_data,rank) 

    // prediction with single test input
    println("single-input prediction made on model:")
    m1.debug_print()
    // will connect to Frovedis master node and get result back
    println("predicted val: " + m1.predict(2,2)) // second user, second product

    // prediction with multiple test inputs
    println("multi-input prediction made on model:")
    m1.debug_print()
    val d = sc.parallelize(Array((1,2), (3,4), (4,1)))
    m1.predict(d).collect.foreach(println)

    // saving model
    m1.save("./out/MyMFModel")
    println("model is saved in: ./out/MyMFModel")

    // loading same model
    val m2 = MatrixFactorizationModel.load("./out/MyMFModel")
    println("loaded model: ./out/MyMFModel:")
    m2.debug_print()

    // recommending users for given product
    println("recommending top 2 users for 2nd product:")
    m2.recommendUsers(2,2).foreach(println) 

    // recommending products for given user
    println("recommending top 2 products for 2nd user:")
    m2.recommendProducts(2,2).foreach(println) 

    FrovedisServer.shut_down()
    sc.stop()
  }
}

File: src/foreign_if/python/examples/als_demo.py
#!/usr/bin/env python

import sys
from frovedis.exrpc.server import FrovedisServer
from frovedis.matrix.crs import FrovedisCRSMatrix
from frovedis.mllib.recommendation import ALS

# initializing the Frovedis server
argvs = sys.argv
argc = len(argvs)
if (argc < 2):
    print ('Please give frovedis_server calling command as the first argument \n(e.g. "mpirun -np 2 -x /opt/nec/nosupport/frovedis/ve/bin/frovedis_server")')
    quit()
FrovedisServer.initialize(argvs[1])

#mat = FrovedisCRSMatrix().load("./input/libSVMFile.txt")
from sklearn.datasets import load_breast_cancer
mat, lbl = load_breast_cancer(return_X_y=True)

# fitting the input matrix on a ALS object
als = ALS(rank=4).fit(mat)

# ids (userId or prodId)  start with 0
print (als.predict([(1,1), (0,1), (2,3), (3,1)]))

# saving the model
als.save("./out/MyMFModel")

# loading the same model
als.load("./out/MyMFModel")

# recommend 2 users for 2nd product
print (als.recommend_users(1,2))

# recommend 2 products for 2nd user
print (als.recommend_products(1,2))

als.release()
FrovedisServer.shut_down()
